### PR TITLE
gh-103000: Optimise dataclasses asdict/astuple for common types

### PR DESCRIPTION
**Pull request by DavidCEllis**
PR for issue #103000 

I'm not attached to `_ATOMIC_TYPES` as the name for the set but 'atomic' matched how they're referred to in `copy` and I didn't have anything more suitable.

* Issue: gh-103000

### CODE REVIEW

File: Lib/dataclasses.py
Comment (AlexWaygood) [inline Lib/dataclasses.py]: Nit: could be a frozenset:

```suggestion
_ATOMIC_TYPES = frozenset({
    # Common JSON Serializable types
    types.NoneType,
    bool,
    int,
    float,
    complex,
    bytes,
    str,
    # Other types that are also unaffected by deepcopy
    types.EllipsisType,
    types.NotImplementedType,
    types.CodeType,
    types.BuiltinFunctionType,
    types.FunctionType,
    type,
    range,
    property,
    # weakref.ref,  # weakref is not currently imported by dataclasses directly
})
```

File: Lib/dataclasses.py
Comment (AlexWaygood) [inline Lib/dataclasses.py]: Add a comment explaining why we have this branch at all (optimising for the common case):

```suggestion
            # Most of the time dict_factory will be dict,
            # so optimize for the common case: 
            result = {}
            for f in fields(obj):
                value = getattr(obj, f.name)
                if type(value) not in _ATOMIC_TYPES:
                    value = _asdict_inner(value, dict_factory)
                result[f.name] = value
            return result
```

File: Lib/dataclasses.py
Comment (AlexWaygood) [inline Lib/dataclasses.py]: I find it so interesting that avoiding the recursive call here appears to save us so much time. Initially when I looked at this PR I wondered if we could do something slightly simpler (https://github.com/python/cpython/compare/main...AlexWaygood:cpython:asdict-faster), but it seems like that is substantially slower on some of your benchmarks!

Comment (carljm) [reply Lib/dataclasses.py]: Python function calls are not fast, and if you have a large list or tuple, this avoids quite a few no-op calls...

It's still not clear to me how much code clarity/simplicity we should sacrifice to micro-optimize these functions. Do you have numbers for the three-way comparison between the status quo, your simpler version, and this version? (It might be good to benchmark with a non-debug build if you aren't already; debug build can skew the relative perf impacts of different code patterns.)

File: Lib/dataclasses.py
Comment (carljm) [inline Lib/dataclasses.py]: I think this comment is overly-focused on deepcopy.

It is important for semantic consistency that all of these types be ones for whom deepcopy is identity, but it is not important that we include here the full list of types for which that is true, and "skipping deepcopy" does not fully describe the optimization that we apply to these types.
```suggestion
# Atomic immutable types which don't require any recursive handling and for which deepcopy
# returns the same object. We can provide a fast-path for these types in asdict and astuple.
```

File: Lib/dataclasses.py
Comment (carljm) [inline Lib/dataclasses.py]: I don't see any value in including this comment.
```suggestion
```

**Review by carljm — COMMENTED** [review]
One idea for a test that could be added here: we could add a test that dataclasses' `_ATOMIC_TYPES` is a subset of deepcopy's. This would protect against a future change to either of those lists breaking the semantic consistency that everything is deep-copied in asdict/astuple.

**Comment by carljm** [comment]
I think this PR should include a news entry for the performance improvement.

File: Lib/dataclasses.py
Comment (AlexWaygood) [reply Lib/dataclasses.py]: I'm using a non-debug, PGO-optimised build already -- I'm learning! ;)

Here's a three-way comparison on my machine, using @DavidCEllis's benchmark (which looks reasonable to me, but I'm no expert on writing benchmarks):

<details>
<summary>`main` branch</summary>

```
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: main
Test Iterations: 10000
List of Int case asdict: 5.71s

Test Iterations: 1000
List of Decimal case asdict: 0.63s

Test Iterations: 1000000
Basic types case asdict: 9.35s
Basic types astuple: 11.84s

Test Iterations: 100000
Opaque types asdict: 7.27s
Opaque types astuple: 7.06s

Test Iterations: 100
Mixed containers asdict: 11.63s
Mixed containers astuple: 10.74s
```

</details>

<details>
<summary>This PR</summary>

```
Running PGUpdate|x64 interpreter...
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: faster_dataclasses_serialize
Test Iterations: 10000
List of Int case asdict: 0.60s

Test Iterations: 1000
List of Decimal case asdict: 0.65s

Test Iterations: 1000000
Basic types case asdict: 1.29s
Basic types astuple: 1.26s

Test Iterations: 100000
Opaque types asdict: 2.00s
Opaque types astuple: 2.15s

Test Iterations: 100
Mixed containers asdict: 6.48s
Mixed containers astuple: 6.37s
```

</details>

<details>
<summary>Alex's "simpler" version</summary>

```
Running PGUpdate|x64 interpreter...
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: asdict-faster
Test Iterations: 10000
List of Int case asdict: 1.31s

Test Iterations: 1000
List of Decimal case asdict: 0.67s

Test Iterations: 1000000
Basic types case asdict: 1.76s
Basic types astuple: 1.67s

Test Iterations: 100000
Opaque types asdict: 2.02s
Opaque types astuple: 5.93s

Test Iterations: 100
Mixed containers asdict: 7.74s
Mixed containers astuple: 6.88s
```

</details>

So: it does seem like both this PR as it currently is and my "simpler" version would be substantially faster than `main`. Also: I believe Python-to-Python calls are meant to be exactly the kind of thing that the faster-CPython team are trying to optimise in the interpreter core itself? So maybe we shouldn't be trying to micro-optimise based on the slowness of Python-to-Python calls _too_ much.

Comment (carljm) [reply Lib/dataclasses.py]: Thanks for doing that benchmarking!

I agree with you that we shouldn't overly micro-optimize based on function calls being slow: it's the responsibility of interpreter optimizations to reduce that cost so we don't have to contort the Python code we write. And the simpler version where we bail out early at the top of `_as*_inner` for atomic types still offers the bulk of the performance win, with much less repetitive code. So I think we should go for the simpler version here, and not sprinkle identical checks against `_ATOMIC_TYPES` throughout every case in these functions.

**Review by carljm — CHANGES_REQUESTED** [review]
Requesting the mentioned inline changes, the suggested test, the news entry, and the change to the simpler version that tests against `_ATOMIC_TYPES` just once at the top of the functions.

Thanks for finding and working on this optimization! Quite an impressive perf win either way.

File: Lib/dataclasses.py
Comment (DavidCEllis) [reply Lib/dataclasses.py]: I actually did this same test with another branch - https://github.com/DavidCEllis/cpython/tree/mid_dataclass_serialize

As the difference in performance was significant I felt it best to do the more annoying but faster variant.

Comment (carljm) [reply Lib/dataclasses.py]: Tbh I'm also curious about the split of the win between the `default_factory` optimization and the `_ATOMIC_TYPES` one. I assume both are significant, and the cost of the `default_factory` change in code complexity is also pretty low, but it would be ideal to evaluate the cost/benefit of each independently.

Comment (AlexWaygood) [reply Lib/dataclasses.py]: > I assume both are significant

Hmm, I tried benchmarking, and I'm not sure the `default_factory` optimisation actually gets us much. Here's benchmark results from https://github.com/python/cpython/compare/main...AlexWaygood:cpython:asdict-faster-no-default-factory-change, which is my "simple" version but with the `default_factory` change removed. It's maybe slightly slower with the `default_factory` change removed, but the difference isn't massive:

<details>

```
Running PGUpdate|x64 interpreter...
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: asdict-faster-no-default-factory-change
Test Iterations: 10000
List of Int case asdict: 1.16s

Test Iterations: 1000
List of Decimal case asdict: 0.67s

Test Iterations: 1000000
Basic types case asdict: 1.87s
Basic types astuple: 1.55s

Test Iterations: 100000
Opaque types asdict: 2.04s
Opaque types astuple: 5.64s

Test Iterations: 100
Mixed containers asdict: 7.94s
Mixed containers astuple: 6.90s
```

</details>

So: +1 to considering the `default_factory` optimisation separately, and postponing it to another PR.

Comment (DavidCEllis) [reply Lib/dataclasses.py]: The `default_factory` optimization is much smaller than the `_ATOMIC_TYPES` one, but was still enough to be consistent and measurable. I can add a performance test that uses a plain subclass for `default_factory`.

Comment (AlexWaygood) [reply Lib/dataclasses.py]: > The `default_factory` optimization is much smaller than the `_ATOMIC_TYPES` one, but was still enough to be consistent and measurable. I can add a performance test that uses a plain subclass for `default_factory`.

Sure, but let's consider each change in isolation: it's a distinct change that has different trade-offs in terms of performance/code readability, and should therefore be a separate PR.

Comment (DavidCEllis) [reply Lib/dataclasses.py]: Updated: https://gist.github.com/DavidCEllis/a2c2ceeeeda2d1ac509fb8877e5fb60d

On some tests it's not enough of the overhead to matter, but it's fairly clear on some of the others (for me at least).

<details>
<summary>Perf result</summary>

```
Dataclasses asdict/astuple speed tests
--------------------------------------
Python v3.12.0alpha6
GIT branch: faster_dataclasses_serialize
Test Iterations: 10000 | dict_factory=dict
List of Int case asdict: 0.55s

Test Iterations: 10000 | dict_factory=CustomDict
List of Int case asdict: 0.56s

Test Iterations: 1000 | dict_factory=dict
List of Decimal case asdict: 0.73s

Test Iterations: 1000 | dict_factory=CustomDict
List of Decimal case asdict: 0.74s

Test Iterations: 1000000 | dict_factory=dict
Basic types case asdict: 1.37s
Basic types astuple: 1.29s

Test Iterations: 1000000 | dict_factory=CustomDict
Basic types case asdict: 1.67s

Test Iterations: 100000 | dict_factory=dict
Opaque types asdict: 2.32s
Opaque types astuple: 2.29s

Test Iterations: 100000 | dict_factory=CustomDict
Opaque types asdict: 2.35s

Test Iterations: 100 | dict_factory=dict
Mixed containers asdict: 2.08s
Mixed containers astuple: 2.00s

Test Iterations: 100 | dict_factory=CustomDict
Mixed containers asdict: 2.41s

```
</details>

Comment (AlexWaygood) [reply Lib/dataclasses.py]: Great! However, adding new benchmarks does not address my argument that the `default_factory` change is a separate optimisation that should be evaluated independently as part of a separate PR

File: Lib/dataclasses.py
Comment (DavidCEllis) [reply Lib/dataclasses.py]: Sorry, I've done this but because I'd already used one of the other suggestions github blocked accepting this change directly.

File: Lib/dataclasses.py
Comment (DavidCEllis) [reply Lib/dataclasses.py]: Sorry, the problem is mostly that due to the interconnected changes I couldn't really offer it as 2 completely separate things. I can make a separate PR but would that refer to the same issue or a new issue too? If so it would be dependant on this one anyway.

Comment (AlexWaygood) [reply Lib/dataclasses.py]: You can attach the followup PR to the same issue (you can probably share one news entry between them :)

File: Lib/dataclasses.py
Comment (AlexWaygood) [reply Lib/dataclasses.py]: Yeah I spotted, no worries :)

File: Lib/dataclasses.py
Comment (DavidCEllis) [reply Lib/dataclasses.py]: I've removed it for now. I think this threading system in github is going to give me a headache though!

File: Lib/dataclasses.py
Comment (DavidCEllis) [reply Lib/dataclasses.py]: This discussion happened in a separate thread - but there will be a separate PR for this as agreed.

**Comment by DavidCEllis** [comment]
A few unresolved things:

* `dict_factory=dict` optimisation will now be done in a separate follow-up PR.
* If I make a test for _ATOMIC_TYPES  is there a specific spot it whould go?
* In the discussion thread Eric posted about getting `copy` to handle/provide the list/set of types
    * I think that's out of scope for this change and conceals *why* these types matter, but maybe worth looking into if it would be useful elsewhere
    * This could always be changed to use such a set if it was provided and if that was the right choice

With regard to the 'simpler' version, while it is a cleaner change I'm not sure it's the right decision to give up on the performance improvement that we can get *now* for the potential that changes to the interpreter might get some of it back in the future?

However I recognise I'm also not the one who has to maintain this and I do understand that it is a bit messier so I'll make the change if that's still requested. (The thumbs up on my comment explaining why I hadn't just done that has left me a bit confused on that front).

**Comment by AlexWaygood** [comment]
I'm sort-of -0 on importing anything private from `copy.py`, whether it's for `dataclasses` itself or a `dataclasses` test. I don't like the idea of coupling the logic of the two modules like that -- and if this change is liable to break unexpectedly in the future when `copy.py`'s implementation changes, we probably shouldn't be making it _anyway_. _But_ I'm not a dataclasses maintainer, so you should probably listen to Carl and Eric over me on this.

Anyway, the test, if you do add it, should go somewhere in `Lib/test/test_dataclasses.py`](https://github.com/python/cpython/blob/main/Lib/test/test_dataclasses.py)

**Comment by carljm** [comment]
> (The thumbs up on my comment explaining why I hadn't just done that has left me a bit confused on that front).

Sorry, my mistake! I've removed the thumbs-up to clarify :) Somehow the first time around I missed the second sentence of that comment, and only saw the first part saying that you had already done the benchmark. I have no explanation for how I managed to miss the second sentence, considering I was actually a bit confused why you were mentioning that you'd already done the benchmark without offering any further conclusions!

> Another way of looking at that is: should we make the code significantly less readable now for an optimisation based on an implementation detail of the current interpreter, that could change at any time? Once the code is checked in, it will be hard to justify changing it in the future for readability reasons, as we highly value code stability in the CPython repo -- we generally only make changes if they fix user-visible bugs or are a meaningful performance improvement. The readability loss could be permanent.

> I would still favour the simpler code.

I agree with this. There is always a balance between readability and performance, and we don't always prefer any detectable improvement in performance at any cost in readability and maintainability. We can get most of the wins here with much nicer code.

> I'm sort-of -0 on importing anything private from copy.py, whether it's for dataclasses itself or a dataclasses test. I don't like the idea of coupling the logic of the two modules like that.

Logically, the coupling is introduced by the optimization itself; the test is just verifying the assumptions of the optimization aren't broken. That said, I think it's very unlikely that `deepcopy` would ever reduce the set of objects that it treats as atomic, and it's more likely they might rearrange the details of how that dispatch happens, so probably the test would be much more likely to fail for spurious and annoying reasons than for real ones. So I've changed my mind: forget the test :)

**Review by AlexWaygood — COMMENTED** [review]
Looks very good to me now, but still needs a news entry!

**Comment by DavidCEllis** [comment]
I think that's all the comments covered?

I'm not sure I'll do a PR for the dict_factory special case, I think that was less significant than skipping the function call overhead? Logically now you'd probably write it as a comprehension (with the inline checks this was much more awkward) and I don't remember that getting as much improvement (perhaps when PEP709 lands).

**Comment by DavidCEllis** [comment]
I have made the requested changes; please review again

**Comment by ericvsmith** [comment]
I'm basically okay with this, although as I mentioned on discuss.python.org I'm concerned about maintaining the list of types here and not in `copy`. This seems to be a generally useful optimization, why enable it only here? Should every module that wants to do something similar maintain it's own copy of `_ATOMIC_TYPES`?

**Comment by DavidCEllis** [comment]
The way I look at it now is that it's not **really** about `deepcopy`.  `deepcopy` ignoring these objects permits us to special case and return them, but if they weren't *useful* types it would be irrelevant. 

The overlap with the types the stdlib `json` can handle and the other potentially common types (complex, bytes) are what made this seem worthwhile to me. The remaining types are mostly there because I didn't have a compelling reason *not* to include them? Including everything at least fit the pattern of 'ignored by deepcopy' rather than being an arbitrary subset that I had chosen (and I definitely didn't think *I* should be the one choosing).

File: Lib/dataclasses.py
Comment (sunmy2019) [inline Lib/dataclasses.py:227]: the name `atomic` might be confusing. How about `primitive`?

**Comment by carljm** [comment]
> I'm concerned about maintaining the list of types here and not in `copy`. This seems to be a generally useful optimization, why enable it only here? Should every module that wants to do something similar maintain it's own copy of `_ATOMIC_TYPES`?

I think this is mostly a dataclasses-specific optimization, where one component of it is avoiding an unnecessary call to deepcopy. A chunk of the win comes from short-circuiting the other type checks done by `_as*_inner` before it even tries deepcopy. Dataclasses, like deepcopy, is implementing a recursive algorithm that should descend only into "non-atomic" types, which is why this optimization can do "double duty" for dataclasses. I doubt it would make sense for other callers of deepcopy who aren't in a similar situation to bother with a pre-check for atomic types.

Given that I don't see other similar uses of deepcopy in the stdlib that would benefit from this, I'm not sure it's worth exposing a new public attribute for this on the copy module? Duplication seems OK here. But I'm also not opposed to making the change in `copy` to expose it.

**Review by carljm — APPROVED** [review]
Code changes LGTM here. Will defer to @ericvsmith on whether he wants `copy` to expose the atomic-types list instead.

**Comment by carljm** [comment]
@ericvsmith considering my comment above, do you still feel that this should be updated to expose a list of atomic types as an attribute of the `copy` module?

File: Misc/NEWS.d/next/Library/2023-03-24-20-49-48.gh-issue-103000.6eVNZI.rst
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-03-24-20-49-48.gh-issue-103000.6eVNZI.rst]: Nit: should wrap documentation (including news entries) to <=80 chars where possible (also a few small wording tweaks):

```suggestion
Improve performance of :func:`dataclasses.astuple` and
:func:`dataclasses.asdict` in cases where the contents are common Python types.
```

**Review by AlexWaygood — APPROVED** [review]
This LGTM, other than my small nit about the NEWS entry.

> Given that I don't see other similar uses of deepcopy in the stdlib that would benefit from this, I'm not sure it's worth exposing a new public attribute for this on the copy module? Duplication seems OK here. But I'm also not opposed to making the change in `copy` to expose it.

I agree with @carljm here. I think a large part of this PR is about avoiding the inner loop in `_asdict_inner` and `_astuple_inner`, rather than any overhead in the `copy` module. It's true that exposing the `_ATOMIC_TYPES` list in the `copy` module _might_ mean that other code could then add similar optimisations more easily. But I think that can be considered separately -- I'd advocate merging this now, and then considering whether to enhance the `copy` module API in a followup issue.

**Pull request merged**
